packages/stitch/tests/eagerReturn.test.ts

```diff
@@ -0,0 +1,185 @@
+import { makeExecutableSchema } from '@graphql-tools/schema';
+import { stitchSchemas } from '@graphql-tools/stitch';
+import { graphql } from 'graphql';
+
+describe('eager returns', () => {
+  let metadataRequests = 0;
+
+  const posts = [
+    { id: '1', productDealId: null, restaurantId: null },
+    { id: '2', productDealId: null, restaurantId: '2' },
+  ];
+
+  const postsSchema = makeExecutableSchema({
+    typeDefs: `
+      type Post {
+        id: ID!
+        productDealId: ID
+        restaurantId: ID
+      }
+
+      type Query {
+        post(id: ID!): Post
+      }
+    `,
+    resolvers: {
+      Query: {
+        post: (root, { id }) => posts.find(post => post.id === id),
+      }
+    }
+  });
+
+  const metadataSchema = makeExecutableSchema({
+    typeDefs: `
+      type ProductDeal {
+        id: ID!
+      }
+      type Restaurant {
+        id: ID!
+      }
+
+      type Post {
+        id: ID!
+        productDeal: ProductDeal
+        restaurant: Restaurant
+      }
+
+      input PostInput {
+        id: ID!
+        productDealId: ID
+        restaurantId: ID
+      }
+
+      type Query {
+        _posts(representations: [PostInput!]!): [Post]!
+      }
+    `,
+    resolvers: {
+      Query: {
+        _posts: (root, { representations }) => {
+          metadataRequests += 1;
+          return representations.map(({ id, productDealId, restaurantId }) => ({
+            id,
+            productDeal: productDealId ? { id: productDealId } : null,
+            restaurant: restaurantId ? { id: restaurantId } : null,
+          }));
+        },
+      }
+    }
+  });
+
+  const getGatewaySchema = (eagerReturn): GraphQLSchema => stitchSchemas({
+    subschemas: [
+      {
+        schema: postsSchema,
+        merge: {
+          Post: {
+            selectionSet: '{ id }',
+            fieldName: 'post',
+            args: ({ id }) => ({ id }),
+          },
+        },
+      },
+      {
+        schema: metadataSchema,
+        merge: {
+          Post: {
+            selectionSet: '{ id }',
+            fields: {
+              productDeal: { selectionSet: '{ productDealId }' },
+              restaurant: { selectionSet: '{ restaurantId }' },
+            },
+            fieldName: '_posts',
+            key: ({ id, productDealId, restaurantId }) => ({ id, productDealId, restaurantId }),
+            argsFromKeys: (representations) => ({ representations }),
+            eagerReturn,
+          },
+        },
+      },
+    ],
+    mergeTypes: true
+  });
+
+  beforeEach(() => {
+    metadataRequests = 0;
+  });
+
+  it('allows early return for empty keys', async () => {
+    const eagerReturn = (key) => null;
```

Comment:
This is the primary usecase here... federation entity keys would write a check along the lines of:

```
(key) => Object.values(key).every(value => value == null) ? null : undefined
```

Reply:
No, I’d be “every”. The representation would only be considered empty if all keys had no data.